File: holo-e2e-performance-tool/src/main/java/com/alibaba/hologres/performace/client/InsertTest.java
package com.alibaba.hologres.performace.client;


import com.alibaba.hologres.client.HoloClient;
import com.alibaba.hologres.client.HoloConfig;
import com.alibaba.hologres.client.Put;

import com.alibaba.hologres.client.impl.ExecutionPool;
import com.alibaba.hologres.client.model.TableSchema;
import com.alibaba.hologres.client.utils.ConfLoader;
import com.alibaba.hologres.client.utils.Metrics;
import com.alibaba.hologres.com.codahale.metrics.Meter;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.util.List;
import java.util.Random;
import java.util.concurrent.atomic.AtomicInteger;
import java.util.concurrent.atomic.AtomicLong;

public class InsertTest extends PutTest {
  public static final Logger LOG = LoggerFactory.getLogger(InsertTest.class);

  private AtomicLong tic = new AtomicLong(0);
  private AtomicInteger singleExecutionPoolJobSize;
  private InsertTestConf insertTestConf = new InsertTestConf();

  @Override
  Runnable buildJob(int id) {
    return new InsertJob(id);
  }

  @Override
  void init() throws Exception {
    ConfLoader.load(confName, "insert.", insertTestConf);
    if (insertTestConf.singleExecutionPool) {
      singleExecutionPoolJobSize = new AtomicInteger(conf.threadSize);
    } else {
      singleExecutionPoolJobSize = new AtomicInteger(0);
    }
  }

  class InsertJob implements Runnable {
    final int id;

    public InsertJob(int id) {
      this.id = id;
    }

    @Override
    public void run() {
      HoloConfig poolConf = new HoloConfig();
      HoloConfig clientConf = new HoloConfig();
      HoloClient client = null;
      HoloClientExecutionPool pool = null;
      try {
        ConfLoader.load(confName, "holoClient.", poolConf);
        ConfLoader.load(confName, "holoClient.", clientConf);
        ConfLoader.load(confName, "pool.", poolConf);

        Meter meter = Metrics.registry().meter(Metrics.METRICS_WRITE_RPS);
        client = new HoloClient(clientConf);
        pool = new HoloClientExecutionPool(poolConf, this.id, insertTestConf.singleExecutionPool);
        pool.setHoloClientPool(client);
        Random rand = new Random();
        TableSchema schema = client.getTableSchema(conf.tableName);
        int i = 0;
        List<String> writeColumns = Util.getWriteColumnsName(conf, schema);
        while (true) {
          long pk = tic.incrementAndGet();
          ++i;
          if(conf.testByTime) {
            if (i % 1000 == 0) {
              if (System.currentTimeMillis() > targetTime) {
                LOG.info("test time reached");
                totalCount.addAndGet(i-1);
                break;
              }
            }
          } else {
            if (pk > conf.rowNumber) {
              LOG.info("insert write : {}", i - 1);
              totalCount.addAndGet(i-1);
              break;
            }
          }
          Put put = newPut(pk, schema, rand, writeColumns, insertTestConf.enableRandomPartialColumn);
          client.put(put);
        }
        client.flush();
      } catch (Exception e) {
        e.printStackTrace();
      } finally {
        if (conf.dumpMemoryStat) {
          try {
            barrier.await();
          } catch (Exception e) {
            e.printStackTrace();
          }
        }
        if(client != null) {
          client.close();
        }
        if (insertTestConf.singleExecutionPool && singleExecutionPoolJobSize.decrementAndGet() == 0 || (!insertTestConf.singleExecutionPool)) {
          if (pool != null) {
            pool.close();
          }
        }
      }
    }
  }


  private Put newPut(long id, TableSchema schema, Random random, List<String> writeColumns, boolean enableRandomPartialColumn) {
    Put put = new Put(schema);
    fillRecord(put.getRecord(), id, schema, random, writeColumns, enableRandomPartialColumn);
    return put;
  }
}

class InsertTestConf {
  public boolean singleExecutionPool = true;
  public boolean enableRandomPartialColumn = false;
}

File: holo-client-py/perf/insert_test.py
"""INSERT mode write test using HoloClient.put() API.

Uses a single HoloClient. Parallelism is controlled by write_parallelism
in the HoloConfig (multi-process workers with shared memory).
"""

from __future__ import annotations

import logging
import random
import time
from typing import Callable

from hologres import HoloClient, Put

from .put_test import PutTest
from .util import get_write_columns

logger = logging.getLogger(__name__)


class InsertTest(PutTest):
    """Write test using HoloClient.put() (batched inserts).

    Parallelism is handled internally by HoloClient via write_parallelism.
    """

    def __init__(self):
        super().__init__()
        self._client: HoloClient | None = None
        self._update_mode = False

    def run(self, conf_name: str, update_mode: bool = False) -> None:
        self._update_mode = update_mode
        super().run(conf_name)

    def init(self) -> None:
        if self._update_mode:
            self.conf.create_table_before_run = False
        self._client = HoloClient(self.holo_config)

    def build_job(self, thread_id: int) -> Callable:
        client = self._client

        def job():
            try:
                schema = client.get_table_schema(self.conf.table_name)
                write_columns = get_write_columns(self.conf, schema)
                rng = random.Random()
                i = 0

                while True:
                    pk = self.tic.increment_and_get()
                    i += 1

                    if i % 1000 == 0 and self.should_stop(i, pk):
                        break

                    put = Put(schema)
                    self.fill_record(put, pk, schema, write_columns, rng)

                    start_ns = time.monotonic_ns()
                    client.put(put)
                    elapsed_ms = (time.monotonic_ns() - start_ns) / 1_000_000
                    self.meter.mark()
                    self.histogram.update(elapsed_ms)

                logger.info("Thread-%d: wrote %d records", thread_id, i)
            except Exception:
                logger.exception("Thread-%d failed", thread_id)

        return job

    def cleanup(self) -> None:
        if self._client is not None:
            try:
                self._client.flush()
                self._client.close()
            except Exception:
                logger.exception("Error closing client")
            self._client = None
